silkworm/capi/silkworm.cpp

```diff
@@ -473,171 +477,267 @@ class BlockProvider {
+    }
+    if (start_block > max_block) {
+        return SILKWORM_INVALID_BLOCK_RANGE;
+    }
+    SignalHandlerGuard signal_guard;
+
+    try {
+        // Wrap MDBX env into an internal *unmanaged* env, i.e. MDBX env is only used but its lifecycle is untouched
+        db::EnvUnmanaged unmanaged_env{mdbx_env};
+        auto txn = db::RWTxnManaged{unmanaged_env};
+        const auto db_path{unmanaged_env.get_path()};
+
+        db::Buffer state_buffer{txn, /*prune_history_threshold=*/0};
+        BoundedBuffer<std::optional<Block>> block_buffer{kMaxBlockBufferSize};
+        BlockProvider block_provider{&block_buffer, unmanaged_env, start_block, max_block};
+        boost::strict_scoped_thread<boost::interrupt_and_join_if_joinable> block_provider_thread(block_provider);
+
+        const size_t max_batch_size{batch_size};
+        auto signal_check_time{std::chrono::steady_clock::now()};
+
+        std::optional<Block> block;
+        BlockNum block_number{start_block};
+        BlockExecutor block_executor{chain_id, write_receipts, write_call_traces, write_change_sets, max_batch_size};
 
-            ++progress.processed_blocks;
-            progress.processed_transactions += block->transactions.size();
-            progress.processed_gas += block->header.gas_used;
-            gas_batch_size += block->header.gas_used;
-
-            // Always flush history for single processed block (no batching)
-            state_buffer.write_history_to_db(write_change_sets);
-
-            // Flush state buffer if we've reached the target batch size
-            if (state_buffer.current_batch_state_size() >= max_batch_size) {
-                log::Info{"[4/12 Execution] Flushing state",  // NOLINT(*-unused-raii)
-                          log_args_for_exec_flush(state_buffer, max_batch_size, block->header.number)};
-                state_buffer.write_state_to_db();
-                // Always save the Execution stage progess when state batch is flushed
-                db::stages::write_stage_progress(*txn, db::stages::kExecutionKey, block->header.number);
-                gas_batch_size = 0;
-                // Commit and renew only in case of internally managed transaction
-                if (!use_external_txn) {
-                    StopWatch sw{/*auto_start=*/true};
-                    txn->commit_and_renew();
-                    const auto [elapsed, _]{sw.stop()};
-                    log::Info("[4/12 Execution] Commit state+history",  // NOLINT(*-unused-raii)
-                              log_args_for_exec_commit(sw.since_start(elapsed), db_path));
+        if (!block_executor.has_known_chain_id()) {
+            return SILKWORM_UNKNOWN_CHAIN_ID;
+        }
+
+        while (block_number <= max_block) {
+            while (state_buffer.current_batch_state_size() < max_batch_size && block_number <= max_block) {
+                block_buffer.pop_back(&block);
+                if (!block || !block.has_value()) {
```

Comment:
It seems a redundant double-check testing exactly the same condition (we already had it but it's redundant)